pkg/controller/jobframework/reconciler.go

```diff
@@ -258,15 +258,17 @@ func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Reques
 	}
 
 	// 2. handle job is finished.
-	if condition, finished := job.Finished(); finished && wl != nil {
+	if condition, finished := job.Finished(); finished {
 		// Execute job finalization logic
 		if err := r.finalizeJob(ctx, job); err != nil {
 			return ctrl.Result{}, err
 		}
 
-		err := workload.UpdateStatus(ctx, r.client, wl, condition.Type, condition.Status, condition.Reason, condition.Message, constants.JobControllerName)
-		if err != nil {
-			log.Error(err, "Updating workload status")
+		if wl != nil && !apimeta.IsStatusConditionTrue(wl.Status.Conditions, kueue.WorkloadFinished) {
+			err := workload.UpdateStatus(ctx, r.client, wl, condition.Type, condition.Status, condition.Reason, condition.Message, constants.JobControllerName)
+			if err != nil {
+				log.Error(err, "Updating workload status")
+			}
```

Comment:
🤔 this might have escaped previous reviews.

If the Workload update fails, we should return the error, so it can be reconciled again (except for NotFound errors).

Another thing that escaped:

Finalizing the job should happen after we update the Workload. Otherwise, the job and the Workload (which is a dependent object) could potentially be deleted before it was transitioned to Finished. This could be fine, but it's better to give complete information before removing the Workload, in case a user is watching for the changes.